src/core/core.js

```diff
@@ -130,31 +130,50 @@ Crafty.fn = Crafty.prototype = {
                 del = rspace;
             }
 
-            //loop over entities
-            for (e in entities) {
-                if (!entities.hasOwnProperty(e)) continue; //skip
-                current = entities[e];
+            if (or) {
+                comps = selector.split(del);
 
-                if (and || or) { //multiple components
-                    comps = selector.split(del);
-                    i = 0;
-                    l = comps.length;
-                    score = 0;
+                filteredCompEnts = {}; // populate object with union of keys from all sets
+                for (i = 0, l = comps.length; i < l; i++) {
+                    compEnts = compEntities[comps[i]];
+                    for (compEnt in compEnts) {
+                        filteredCompEnts[compEnt] = +compEnt; //convert to int
+                    }
+                }
 
-                    for (; i < l; i++) //loop over components
-                        if (current.__c[comps[i]]) score++; //if component exists add to score
+                for (compEnt in filteredCompEnts) { // add set elements to result
+                    this[elem++] = filteredCompEnts[compEnt];
+                }
+            } else if (and) {
+                comps = selector.split(del);
+
+                filteredCompEnts = {}; // populate initial object with keys in common from first two sets
+                compEnts = compEntities[comps[0]];
+                compEnts2 = compEntities[comps[1]];
+                for (compEnt in compEnts) {
+                    if (compEnts2[compEnt] !== undefined)
+                        filteredCompEnts[compEnt] = +compEnt; //convert to int
+                }
 
-                        //if anded comps and has all OR ored comps and at least 1
-                    if (and && score === l || or && score > 0) this[elem++] = +e;
+                for (i = 2, l = comps.length; i < l; i++) { // strip initial object keys not in following sets
```

Comment:
I was not trying to optimize it for 2 component names per se :)   
From an algorithmic standpoint I guess the fastest thing to do is create a temporary map holding all ids for the first compName and then remove all ids that are not in common with the map for another compName.   
As we have to iterate all ids once to copy them over to the temporary map at the start, why not use this iteration to do some work already?